### ISSUE TITLE
WhatsApp-Business Channel

### ISSUE THREAD

**Issue by Mirtaaa** [zammad/zammad#5010]
### Summary
Zammad as a helpdesk software has a lot of communication channels built in. They are used to keep in touch with customers and, for sure, to offer the best service that is possible.
Currently, Zammad lacks the possibility to use WhatsApp as one way of communicating with customers, which is very popular nowadays.

### Use Cases
1.  As a customer, I want to open tickets by chatting with a company via WhatsApp.
2. As a customer, I want to receive information on my tickets by a WhatsApp message.
3. As an admin, I want to configure multiple WhatsApp Business Numbers and route them to different groups.
4. As an admin, I want to configure a trigger that sends out a WhatsApp message in case of an emergency call by a VIP customer (even outside of defined business hours).

<details>
<summary>Additional use cases for a later iteration</summary>

- As an agent, I want to be able to send a WhatsApp message within an existing ticket to get in touch with one of my colleagues. In WhatsApp, it's possible to initiate the connection with a target user in certain situations (advanced subscription plan). This may make agent-to-agent use cases viable, at least from a technical perspective. It's an entirely different matter if this is something that makes sense for Zammad as a ticket system.
- Creation of outbound tickets via WhatsApp channel. As an agent, I want to be able to get in touch with customers directly via WhatsApp.

_Currently, it's not possible to say for sure if all use cases are technically feasible. We need to dig deeper, but one thing is clear: The initial use case for responding to a customer-initiated communication within the allowed time window seems possible. Everything else is still just theoretically possible._

</details>

### ACCEPTANCE CRITERIA
- Create a ticket when a customer sends a Message via WhatsApp Business Channel
- Usage of multiple phone numbers
- Normal ticket communication where the agent can continue the conversation with the customer
- Zammad must take care of sending reminders for all WhatsApp tickets that are currently open to avoid losing the conversation with the customer (see https://internalnote.com/breaking-the-24h-rule-for-whatsapp-in-zendesk/).
- It must be possible to send images via WhatsApp to Zammad.

#### Ideas for later iteration
- Scheduler/triggers should work with incoming WhatsApp tickets (Limitations: Notification only possible in 24 Windows without usage of the WhatsApp template)
- ~Creation of outbound tickets and answers without an existing WhatsApp conversation via WhatsApp channel~ (to make this possible we would need templates approved by Meta and the other side needs to give the agreement first that you are allowed to contact him

### Known Requirements and Limitations
- Verified Meta Business Account (Business App is distinct from personal Mobile App, much stricter parameters but more features are available)
- registered app in Meta for Developers (Display name has to match the registered business name, otherwise it may be rejected!)
- A Business Manager contact is required
- WhatsApp integration must be enabled
- The phone number used needs to be able to receive international phone calls/SMS (Meta requirement)
- A phone number can't be connected with the WhatsApp for Business mobile app and the WhatsApp Business API at the same time (Meta limitation). Only a fresh phone number can be used with the WhatsApp Business API (not tied to the mobile app).
- If your phone number is currently in use with the WhatsApp for Business mobile app, you will need to follow Meta's [process](https://developers.facebook.com/docs/whatsapp/on-premises/get-started/migrate-existing-whatsapp-number-to-a-business-account) to remove the account.
- A phone number that is connected to the WhatsApp Business API is not usable for incoming WhatsApp calls anymore.
- The display name needs to match your business name/branding, otherwise, WhatsApp will reject it.
- A Meta Business Manager for the company is needed.
- Need to check the details of how to start a conversation with a user directly (e.g. agent -> customer, agent -> agent)
- Need to check if notifications via WhatsApp are possible
- Possible to initiate communication with a customer only with a pre-approved message template.

**Pricing/different kinds of messages**
This is a topic that should maybe be explained in the documentation or at least some references should be added to find some information.

**24h window to reply**
WhatsApp for example has a 24-hour rule which prevents business integrations from replying to customers that haven't interacted with your business for over 24 hours. The 24-hour window for reply, outside's not allowed; possibility to assist the customer to extend this window on-demand when the limit is approaching.

Example: If a customer sends a ticket at 10 AM, you need to make sure your agent replies AND you get a reply from the customer before 10 AM the next day. If those 24h pass without a customer reply to your agents' remark, you can't send any reminders or more info to that customer anymore.

----
**Additional information from other sources:**
https://developers.facebook.com/docs/whatsapp/overview/getting-opt-in/?locale=de_DE
https://business.whatsapp.com/policy
https://developers.facebook.com/docs/whatsapp/conversation-types#faq_715120295540784
https://developers.facebook.com/docs/messenger-platform/policy/policy-overview/

<details>
<summary>Sources regarding Creation of Outbound Tickets and Templates </summary>

**About starting a conversation from Zammad's side**

https://developers.facebook.com/docs/whatsapp/cloud-api/guides/send-message-templates?locale=en_US

> WhatsApp message templates are specific message formats that businesses use to send out notifications or customer care messages to people who have opted into notifications. Messages can include appointment reminders, shipping information, issue resolution, or payment updates.

> Before sending a template message, you need to create a template. See [Create Message Templates for Your WhatsApp Business Account](https://www.facebook.com/business/help/2055875911147364) for more information. If your account is not verified yet, you can use one of our [pre-approved templates](https://www.facebook.com/business/help/722393685250070).

https://developers.facebook.com/docs/whatsapp/message-templates/guidelines

> Templates are used in template messages to open marketing, utility, and authentication conversations with customers. Unlike free-form messages, template messages are the only type of message that can be sent to customers who have yet to message you, or who have not sent you a message in the last 24 hours.

> Templates must be approved before they can be sent in template messages. In addition, templates may be disabled automatically based on customer feedback. Once disabled, a template cannot be sent in a template message until its quality rating has improved or it no longer violates our business or commerce policies.
</details>

#### Admin - [Link-Figma](https://www.figma.com/file/P3sOAIFK20RQVrx7TdURXA/Zammad-Desktop-App?type=design&node-id=31%3A6209&mode=design&t=6zVMtRutyZ2MkH7J-1)

  <a href="https://github.com/zammad/zammad/assets/6049445/455b21cc-c9a2-4c95-8823-d76e6712c597" target="_blank"><img width="192" alt="01WithoutAccounts" src="https://github.com/zammad/zammad/assets/6049445/455b21cc-c9a2-4c95-8823-d76e6712c597"></a> <a href="https://github.com/zammad/zammad/assets/6049445/a74b4d6e-1ab9-4191-9716-7473d941c87c" target="_blank"><img width="192" alt="02AccountModalStep1" src="https://github.com/zammad/zammad/assets/6049445/a74b4d6e-1ab9-4191-9716-7473d941c87c"></a> <a href="https://github.com/zammad/zammad/assets/6049445/0504b7c0-464d-438c-b262-663190adeac9" target="_blank"><img width="192" alt="02AccountModalStep2" src="https://github.com/zammad/zammad/assets/6049445/0504b7c0-464d-438c-b262-663190adeac9"></a> <a href="https://github.com/zammad/zammad/assets/6049445/ccab30c5-2788-48f8-966e-d2b7757d4906" target="_blank"><img width="192" alt="02AccountModalStep3" src="https://github.com/zammad/zammad/assets/6049445/ccab30c5-2788-48f8-966e-d2b7757d4906"></a> <a href="https://github.com/zammad/zammad/assets/6049445/7917be53-dac2-4f7b-8a23-078701855553" target="_blank"><img width="192" alt="03WithAccount" src="https://github.com/zammad/zammad/assets/6049445/7917be53-dac2-4f7b-8a23-078701855553"></a> <a href="https://github.com/zammad/zammad/assets/6049445/4728f116-4630-4aaf-b211-9e0ab8b015fb" target="_blank"><img width="192" alt="04AccountModalWithAccountStep1" src="https://github.com/zammad/zammad/assets/6049445/4728f116-4630-4aaf-b211-9e0ab8b015fb"></a> <a href="https://github.com/zammad/zammad/assets/6049445/195563fd-9fd1-417d-a600-f755920e43fc" target="_blank"><img width="192" alt="04AccountModalWithAccountStep2" src="https://github.com/zammad/zammad/assets/6049445/195563fd-9fd1-417d-a600-f755920e43fc"></a> <a href="https://github.com/zammad/zammad/assets/6049445/9f92b36c-2ed7-4de8-a483-0266e2ba74da" target="_blank"><img width="192" alt="04AccountModalWithAccountStep3" src="https://github.com/zammad/zammad/assets/6049445/9f92b36c-2ed7-4de8-a483-0266e2ba74da"></a> <a href="https://github.com/zammad/zammad/assets/6049445/a3f72c71-9474-4b91-aa07-0d62efc03dad" target="_blank"><img width="192" alt="05WithDisabled" src="https://github.com/zammad/zammad/assets/6049445/a3f72c71-9474-4b91-aa07-0d62efc03dad"></a> <a href="https://github.com/zammad/zammad/assets/6049445/9874d225-635b-464b-9f95-f63e0158c5e1" target="_blank"><img width="192" alt="06WithConfirmation" src="https://github.com/zammad/zammad/assets/6049445/9874d225-635b-464b-9f95-f63e0158c5e1"></a>

#### Agent - [Figma-Link](https://www.figma.com/file/P3sOAIFK20RQVrx7TdURXA/Zammad-Desktop-App?type=design&node-id=42%3A7342&mode=design&t=6zVMtRutyZ2MkH7J-1)

  <a href="https://github.com/zammad/zammad/assets/6049445/adcf131a-a63a-47cc-8287-455158e533eb" target="_blank"><img width="192" alt="07IncomingTicket" src="https://github.com/zammad/zammad/assets/6049445/adcf131a-a63a-47cc-8287-455158e533eb"></a> <a href="https://github.com/zammad/zammad/assets/6049445/9e525526-ebc3-42ce-9f08-efe1a27005a0" target="_blank"><img width="192" alt="08Reply" src="https://github.com/zammad/zammad/assets/6049445/9e525526-ebc3-42ce-9f08-efe1a27005a0"></a> <a href="https://github.com/zammad/zammad/assets/6049445/314103d1-7aa6-4dec-9ca9-d88133b636fc" target="_blank"><img width="192" alt="09Headers" src="https://github.com/zammad/zammad/assets/6049445/314103d1-7aa6-4dec-9ca9-d88133b636fc"></a> <a href="https://github.com/zammad/zammad/assets/6049445/e0d1c37d-4eb0-4274-9000-f0a9732d0ea4" target="_blank"><img width="192" alt="10TwoArticles" src="https://github.com/zammad/zammad/assets/6049445/e0d1c37d-4eb0-4274-9000-f0a9732d0ea4"></a> <a href="https://github.com/zammad/zammad/assets/6049445/e93c50da-786e-4abf-83f2-f678a7f9f25d" target="_blank"><img width="192" alt="11ThreeArticles" src="https://github.com/zammad/zammad/assets/6049445/e93c50da-786e-4abf-83f2-f678a7f9f25d"></a> <a href="https://github.com/zammad/zammad/assets/6049445/219db031-4efc-4f27-afad-8b0e1be56245" target="_blank"><img width="192" alt="13Closed" src="https://github.com/zammad/zammad/assets/6049445/219db031-4efc-4f27-afad-8b0e1be56245"></a> <a href="https://github.com/zammad/zammad/assets/6049445/036bbb72-7765-417d-8846-b9307590146a" target="_blank"><img width="192" alt="1224hWindowClosed" src="https://github.com/zammad/zammad/assets/6049445/036bbb72-7765-417d-8846-b9307590146a"></a>

#### Gem or completely own implementation?
Maybe this gem can be used?
https://github.com/facebook/facebook-ruby-business-sdk

#### Can we send images/attachments from our side?

Check how images can be sent from our side (related to the WhatsApp-API) and if it's matching with our current editor situation.
Maybe we need a different kind of WhatsApp article type?
In the end, a decision is needed, if we implement the Zamamd->WhatsApp direction for images/attachments (the other way is no problem).

#### Detection when a new ticket needs to be created or a message should be added in an existing one

Here we need to find a good solution. Maybe check also already existing channels, but maybe here is no special logic behind.

```[tasklist]
### Tasks
- [ ] https://github.com/zammad/coordination-scrum/issues/116
- [ ] https://github.com/zammad/coordination-scrum/issues/117
- [ ] https://github.com/zammad/coordination-scrum/issues/118
- [ ] https://github.com/zammad/coordination-scrum/issues/120
- [ ] https://github.com/zammad/coordination-scrum/issues/119
- [ ] https://github.com/zammad/coordination-scrum/issues/121
- [ ] https://github.com/zammad/coordination-scrum/issues/124
- [ ] https://github.com/zammad/coordination-scrum/issues/123
- [ ] https://github.com/zammad/coordination-scrum/issues/126
- [ ] https://github.com/zammad/coordination-scrum/issues/127
- [ ] https://github.com/zammad/coordination-scrum/issues/128
- [ ] https://github.com/zammad/coordination-scrum/issues/129
- [ ] https://github.com/zammad/coordination-scrum/issues/130
- [ ] https://github.com/zammad/coordination-scrum/issues/131
- [ ] https://github.com/zammad/coordination-scrum/issues/132
```
